### PR TITLE
feat(oidc): user + other improvements

### PR DESCRIPTION
**Pull request by rbordeanu**
What's new:

- strong(er) typing (i.e. all functions that return a value should be typed)
- new user info related interfaces
- special treatment of /user endpoint
- add config options for custom mappings of user id and user name
- prefer `name` to `username` for fallback user name mapping
- add config option for post logout uri (we sometimes want to go to `/` for example, and not always to `/login`)

### CODE REVIEW

File: libs/oidc/src/controllers/auth.controller.ts [hunk 1]
```diff
@@ -3,15 +3,24 @@ import { Response, Request } from 'express';
...
+  @Public()
   @Get('/user')
-  user(@Req() req) {
-    return req.user.userinfo;
+  user(@Req() req: Request): UserInfo | UserinfoResponse {
+    if (req.user) {
+      return req.user['userinfo'];
+    }
+
+    return {
```
Comment (FrancineO) [inline libs/oidc/src/controllers/auth.controller.ts:21]: If not authenticated, the route should be protected by token guard or user middleware

File: libs/oidc/src/controllers/auth.controller.ts [hunk 2]
```diff
@@ -33,13 +42,13 @@ export class AuthController {
   }
 
   @Get('/refresh-token')
-  refreshTokens(@Req() req, @Res() res, @Next() next: Function) {
+  refreshTokens(@Req() req: Request, @Res() res, @Next() next: Function) {
```
Comment (FrancineO) [inline libs/oidc/src/controllers/auth.controller.ts:45]:
```suggestion
  refreshTokens(@Req() req: Request, @Res() res: Response, @Next() next: Function) {
```

File: libs/oidc/src/oidc.module.ts [hunk 3]
```diff
@@ -32,7 +32,10 @@ import { TenantSwitchController } from './controllers';
 })
 export class OidcModule implements NestModule {
   configure(consumer: MiddlewareConsumer) {
-    consumer.apply(UserMiddleware).forRoutes({ path: '*', method: RequestMethod.ALL });
+    consumer
+      .apply(UserMiddleware)
+      .exclude({ path: '/user', method: RequestMethod.GET })
```
Comment (FrancineO) [inline libs/oidc/src/oidc.module.ts:37]: Why do we need to exclude this route?

File: libs/oidc/src/interfaces/oidc-module-options.interface.ts [hunk 4]
```diff
@@ -1,15 +1,18 @@
 import { ModuleMetadata, Type } from '@nestjs/common/interfaces';
 import { AuthorizationParameters, ClientMetadata, HttpOptions } from 'openid-client';
+import { UserInfoMapping } from './user.interface';
 
 export type OidcModuleOptions = {
   origin: string;
   authParams: AuthorizationParameters;
   redirectUriLogout?: string;
+  postLogoutUri?: string;
```
Comment (FrancineO) [inline libs/oidc/src/interfaces/oidc-module-options.interface.ts:9]: Could we name it `postLogoutRedirectUri` :)

Comment (FrancineO) [reply libs/oidc/src/interfaces/oidc-module-options.interface.ts:9]: + Don't forget the doc please

Comment (rbordeanu) [reply libs/oidc/src/interfaces/oidc-module-options.interface.ts]: Done.

File: libs/oidc/src/utils/user-info.ts [hunk 5]
```diff
@@ -19,20 +24,26 @@ export async function getUserInfo(token: string, oidcService: OidcService, idpKe
...
+  oidcService: OidcService,
+  idpKey: string,
+): Promise<UserInfo | UserinfoResponse> {
   try {
     return await oidcService.idpInfos[idpKey].client.userinfo(token);
   } catch (err) {
     const msg = `Error accessing user information`;
     logger.error(msg);
-    return userInfo(token, 'sub');
+    return userInfo(token, { id: 'sub', username: 'sub' });
```
Comment (FrancineO) [inline libs/oidc/src/utils/user-info.ts:37]: Sometimes user info remotes returns more than just the `sub`, maybe we should remove the mapping here and add  `username: identity[usernameParameter] || identity.username || identity.name || identity.sub` in `userInfo function`

Comment (FrancineO) [inline libs/oidc/src/utils/user-info.ts:37]:
```suggestion
    return userInfo(token);
```

File: libs/oidc/src/utils/user-info.ts [hunk 6]
```diff
@@ -19,20 +24,26 @@ export async function getUserInfo(token: string, oidcService: OidcService, idpKe
...
 }
 
-function userInfo(token: string, usernameParameter?: string) {
+function userInfo(token: string, userInfoMapping?: UserInfoMapping): UserInfo {
   const identity: any = JWT.decode(token);
+  const { id, username } = userInfoMapping || {};
   return {
-    username: identity[usernameParameter] || identity.username || identity.name,
+    id: identity[id] || identity.sub,
+    username: identity[username] || identity.name || identity.username,
```
Comment (FrancineO) [inline libs/oidc/src/utils/user-info.ts:46]:
```suggestion
    username: identity[username] || identity.name || identity.username || identity.sub,
```

**Review by FrancineO — APPROVED** [review]

**Pull request merged**